### PR TITLE
backends.ldap: add timeout and paginated results

### PR DESCRIPTION
**Pull request by qbey**
I propose to add two options:
 - ``DATABASES['ldap']['CONNECTION_OPTIONS']['TIMEOUT'] = 2``
   which allows to define the database connection timeout
   (not a big deal).
 - ``DATABASES['ldap']['CONNECTION_OPTIONS']['BATCH_SIZE'] = 500``
   which allows to split requests to the LDAP, thus allowing
   to fetch more data in a single request.

This second option uses `search_ext_s` if defined, and continues
to use previous `search_s` if not. That's why I renamed
`DatabaseWrapper.search_s` with `batched_search_s`.

### CODE REVIEW

**Comment by qbey** [comment]
I might need some help to write few tests :)

File: ldapdb/backends/ldap/base.py [hunk 1]
```diff
@@ -208,7 +210,12 @@ def get_new_connection(self, conn_params):
 
         options = conn_params['options']
         for opt, value in options.items():
-            connection.set_option(opt, value)
+            if opt.upper() == 'TIMEOUT':
+                connection.timeout = int(value)
+            elif opt.upper() == 'BATCH_SIZE':
```
Comment (rbarrois) [inline ldapdb/backends/ldap/base.py]: Not sure about the ``.upper()`` part — is is a common convention in Django?

File: ldapdb/backends/ldap/base.py [hunk 2]
```diff
@@ -180,6 +181,7 @@ def __init__(self, *args, **kwargs):
         self.validation = DatabaseValidation(self)
         self.settings_dict['SUPPORTS_TRANSACTIONS'] = True
         self.autocommit = True
+        self.batch_size = None
```
Comment (rbarrois) [inline ldapdb/backends/ldap/base.py]: Should we set the ``batch_size`` per-project, per-model or per-query?

File: ldapdb/backends/ldap/base.py [hunk 3]
```diff
@@ -252,13 +259,46 @@ def rename_s(self, dn, newrdn):
         cursor = self._cursor()
         return cursor.connection.rename_s(dn, newrdn)
 
-    def search_s(self, base, scope, filterstr='(objectClass=*)',
-                 attrlist=None):
+    def batched_search_s(self, base, scope, filterstr='(objectClass=*)',
+                         attrlist=None):
```
Comment (rbarrois) [inline ldapdb/backends/ldap/base.py]: I'd rather keep the old name, and pass the ``batch_size`` as a paremeter to this query.

File: ldapdb/backends/ldap/base.py [hunk 4]
```diff
@@ -252,13 +259,46 @@ def rename_s(self, dn, newrdn):
...
+                    serverctrls=[ldap_control],
+                    timeout=timeout,
+                )
+
+                _res_type, results, _res_msgid, server_controls = cursor.connection.result3(msgid, timeout=timeout)
+                page_controls = [ctrl for ctrl in server_controls if ctrl.controlType == ldap.CONTROL_PAGEDRESULTS]
+                if not page_controls:
+                    raise self.Database.DatabaseError("Server doesn't support paging (RFC2696)")
+
+                output += [(dn, attrs) for dn, attrs in results if dn is not None]
```
Comment (rbarrois) [inline ldapdb/backends/ldap/base.py]: Maybe explain why we skip entries without a ``dn``?

File: ldapdb/backends/ldap/base.py (hunk 2 above)
Comment (qbey) [inline ldapdb/backends/ldap/base.py]: Actually, I believe the `batch_size` is database dependent, so it's only defined in the Django's `settings.DATABASES`. At the first place I wondering if a per-query `batch_size` would be helpfull, but as long as the database does not support it, or it doesn't allow enough results (`ldap.SIZELIMIT_EXCEEDED`), we might not give the choice to override it in a query.
It's different for the `TIMEOUT` argument I guess.

File: ldapdb/backends/ldap/base.py (hunk 1 above)
Comment (qbey) [inline ldapdb/backends/ldap/base.py]: You're right, I have to fix this, Django uses lower snake case for the `CONNECTION_OPTIONS` keys.

**Pull request merged**